crates/iceberg/src/spec/arrow.rs

```diff
@@ -0,0 +1,553 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::error::Result;
+use crate::spec::{
+    ListType, MapType, NestedField, NestedFieldRef, PrimitiveType, Schema, StructType, Type,
+};
+use crate::{Error, ErrorKind};
+use arrow_schema::{DataType, Field, Fields, Schema as ArrowSchema, TimeUnit};
+use std::sync::Arc;
+
+/// A post order arrow schema visitor.
+///
+/// For order of methods called, please refer to [`visit_schema`].
+pub trait ArrowSchemaVisitor {
+    /// Return type of this visitor.
+    type T;
+    type U;
+
+    /// Called before struct/list/map field.
+    fn before_field(&mut self, _field: &Field) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called after struct/list/map field.
+    fn after_field(&mut self, _field: &Field) -> Result<()> {
+        Ok(())
+    }
+
```

Comment:
Hmm okay. I don't have added them because the python implementation in the ticket is https://github.com/apache/iceberg-python/blob/6c8ea0effac0942ad4e880e5eef627473a354040/pyiceberg/io/pyarrow.py#L604.

This `PyArrowSchemaVisitor` only has a pair of `before_field` and `after_field` for struct/map/list.

Reply:
Sorry, I recently pulled the latest commits in pyiceberg and they added this.

Reply:
No problem. I will add these functions.